### ISSUE TITLE
Opentracing tracers can't override envoy's sampling decision

### ISSUE THREAD

**Issue by cgilmour** [envoyproxy/envoy#9076]
This problem has affected a handful of Datadog users that are using Envoy directly or indirectly via istio, and the X-Request-Id header is missing or its value isn't a format that envoy expects.

Note: this is similar to #7624 and #7718, but would not be solved by a fix for that.

With zipkin tracing, when a request comes in and has both `x-b3-sampled` and `x-request-id` headers, it will use the value of `x-b3-sampled` for its sampling decision, and fall back to envoy's decision based on `x-request-id`. [code](https://github.com/envoyproxy/envoy/blob/648cfc3b6c1410d8b27aa33b6fc0392ab230d601/source/extensions/tracers/zipkin/span_context_extractor.cc#L37-L73)
With opentracing-based tracers, that doesn't happen - it only uses envoy's sampling decision. [code](https://github.com/envoyproxy/envoy/blob/648cfc3b6c1410d8b27aa33b6fc0392ab230d601/source/extensions/tracers/common/ot/opentracing_driver_impl.cc#L194-L197)

For the datadog tracer, this causes incomplete traces - spans are dropped even when its tracing headers said to sample it.

I think the correct fix for this is making opentracing-based tracers have the same behavior as the zipkin tracer.

As usual, happy to work on a fix for this.
I suspect it'll require extending the [OpenTracingDriver](https://github.com/envoyproxy/envoy/blob/648cfc3b6c1410d8b27aa33b6fc0392ab230d601/source/extensions/tracers/common/ot/opentracing_driver_impl.h#L56) and modifying the drivers that use it.

**Comment by junr03** [comment]
Sounds reasonable. If this is something that would build on top of #7624 could you please collaborate with the person working on that issue?

**Comment by cgilmour** [comment]
Still intending to work on this in the near future.

**Comment by cgilmour** [comment]
I think this would either not be affected by #7624, or would be affected equally compared to other tracers.

I've implemented this locally, and since it's affecting the common OpenTracingDriver, would appreciate a second opinion on the approach.

At the moment, the opentracing API doesn't have a way of asking the tracer if a span / spancontext represents a trace with a sampling decision. This could potentially be added to opentracing 1.6.x, but that release had breaking changes (new required methods) and envoy remains on 1.5.1 for now.
(All tracers would need to ship a release supporting 1.6.x before envoy could update to that.)

So an easier approach for now is adding a virtual `sampled` method to the OpenTracingDriver with a default implementation. [here](https://github.com/DataDog/envoy/blob/cgilmour/sampling-decision/source/extensions/tracers/common/ot/opentracing_driver_impl.h#L78-L80)
Then specific drivers can optionally override it. [here](https://github.com/DataDog/envoy/blob/cgilmour/sampling-decision/source/extensions/tracers/datadog/datadog_tracer_impl.cc#L59-L61)
The sampling decision then becomes a combination of the tracer's knowledge and the decision envoy wanted to make.
```diff
-  if (!tracing_decision.traced) {
+  if (!sampled(*parent_span_ctx, tracing_decision)) {
```

If there's a move to opentracing 1.6.x in the future, and it supports a direct way of checking for a propagated sampling decision, it'd be a small change to switch to that.

Assuming that sounds fine, I'll complete the code changes and submit a PR.

**Comment by mattklein123** [comment]
> Assuming that sounds fine, I'll complete the code changes and submit a PR.

Yup SGTM.

**Comment by cgilmour** [comment]
Seems I was wrong on this one. For propagated traces, it's simple enough to just ignore the `sampling_priority` tag that gets passed when spans are created.

The remaining part of the problem (when traces are being initiated) is essentially the same as all the duplicates of #7624 where different behavior is desired.
And a PR for it has just landed at #10018
